src/HealthGPS/kevin_hall_model.cpp

```diff
@@ -114,6 +139,89 @@ void KevinHallModel::update_risk_factors(RuntimeContext &context) {
     }
 }
 
+void KevinHallModel::initialise_sector(RuntimeContext &context, Person &person) const {
+
+    // Get rural prevalence for age group and sex.
+    double prevalence;
+    if (person.age < 18) {
+        prevalence = rural_prevalence_.at("Under18"_id).at(person.gender);
+    } else {
+        prevalence = rural_prevalence_.at("Over18"_id).at(person.gender);
+    }
+
+    // Sample the person's sector.
+    double rand = context.random().next_double();
+    auto sector = rand < prevalence ? core::Sector::rural : core::Sector::urban;
+    person.sector = sector;
+}
+
+void KevinHallModel::update_sector(RuntimeContext &context, Person &person) const {
+
+    // Only update rural sector 18 year olds.
+    if ((person.age != 18) || (person.sector != core::Sector::rural)) {
+        return;
+    }
+
+    // Get rural prevalence for age group and sex.
+    double prevalence_under18 = rural_prevalence_.at("Under18"_id).at(person.gender);
+    double prevalence_over18 = rural_prevalence_.at("Over18"_id).at(person.gender);
+
+    // Compute random rural to urban transition.
+    double rand = context.random().next_double();
+    double p_rural_to_urban = 1.0 - prevalence_over18 / prevalence_under18;
+    if (rand < p_rural_to_urban) {
+        person.sector = core::Sector::urban;
+    }
+}
+
+void KevinHallModel::initialise_income(RuntimeContext &context, Person &person) const {
+
+    // Compute logits for each income category.
+    auto logits = std::vector<double>(income_models_.size());
+    for (size_t i = 0; i < income_models_.size(); i++) {
+        logits[i] = income_models_[i].intercept;
+        for (const auto &[factor_name, coefficient] : income_models_[i].coefficients) {
+            logits[i] += coefficient * person.get_risk_factor_value(factor_name);
+        }
+    }
+
+    // Compute softmax probabilities for each income category.
+    auto e_logits = std::vector<double>(income_models_.size());
+    double e_logits_sum = 0.0;
+    for (size_t i = 0; i < income_models_.size(); i++) {
+        e_logits[i] = exp(logits[i]);
+        e_logits_sum += e_logits[i];
+    }
+
+    // Compute income category probabilities.
+    auto probabilities = std::vector<double>(income_models_.size());
+    for (size_t i = 0; i < income_models_.size(); i++) {
+        probabilities[i] = e_logits[i] / e_logits_sum;
+    }
```

Comment:
These could also use `std::ranges::transform`, though the way you've done it is fine too!

One thing I would say is that generally it's better to append to `std::vector`s iteratively rather than creating them with a particular size upfront (it means you can skip the zeroing of the memory). You can use the `reserve()` method to hint at how many elements it will have eventually.

Reply:
Fair enough with reserve, but I think I prefer the readability of for loops for such simple ops.